File: ios/maimaid/Services/OtogamePlaylogModels.swift
import Foundation

struct OtogamePlaylogResponse: Decodable, Sendable {
    let code: String
    let message: String
    let data: OtogamePlaylogPage
}

struct OtogameRatingResponse: Decodable, Sendable {
    let data: OtogameRatingData
}

struct OtogameRatingData: Decodable, Sendable {
    let ratingList: [OtogameRatingEntry]
    let newRatingList: [OtogameRatingEntry]

    private enum CodingKeys: String, CodingKey {
        case ratingList = "rating_list"
        case newRatingList = "new_rating_list"
    }
}

struct OtogameRatingEntry: Decodable, Sendable {
    let music: OtogameMusic
    let levelInfo: OtogameLevelInfo
    let achievement: Int64
    let comboStatus: Int

    private enum CodingKeys: String, CodingKey {
        case music
        case levelInfo = "level_info"
        case achievement
        case comboStatus = "combo_status"
    }
}

struct OtogamePlaylogPage: Decodable, Sendable {
    let data: [OtogamePlaylog]
    let pagination: OtogamePagination
}

struct OtogamePagination: Decodable, Sendable {
    let page: Int
    let perPage: Int
    let totalPage: Int

    private enum CodingKeys: String, CodingKey {
        case page
        case perPage
        case perPageSnake = "per_page"
        case totalPage
        case totalPageSnake = "total_page"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        page = try container.decodeIfPresent(Int.self, forKey: .page) ?? 1
        perPage = try container.decodeIfPresent(Int.self, forKey: .perPageSnake)
            ?? container.decodeIfPresent(Int.self, forKey: .perPage)
            ?? 0
        totalPage = try container.decodeIfPresent(Int.self, forKey: .totalPageSnake)
            ?? container.decodeIfPresent(Int.self, forKey: .totalPage)
            ?? 1
    }
}

struct OtogamePlaylog: Decodable, Sendable {
    let music: OtogameMusic
    let difficulty: Int
    let levelInfo: OtogameLevelInfo
    let trackNumber: Int
    let playDate: Int64
    let achievement: Int64
    let scoreRank: Int
    let deluxeScore: Int
    let comboStatus: Int
    let syncStatus: Int

    private enum CodingKeys: String, CodingKey {
        case music
        case difficulty
        case levelInfo = "level_info"
        case trackNumber = "track_no"
        case playDate = "play_date"
        case achievement
        case scoreRank = "score_rank"
        case deluxeScore = "deluxe_score"
        case comboStatus = "combo_status"
        case syncStatus = "sync_status"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        music = try container.decode(OtogameMusic.self, forKey: .music)
        difficulty = try container.decodeIfPresent(Int.self, forKey: .difficulty) ?? -1
        levelInfo = try container.decodeIfPresent(OtogameLevelInfo.self, forKey: .levelInfo) ?? .empty
        trackNumber = try container.decodeIfPresent(Int.self, forKey: .trackNumber) ?? 0
        playDate = try container.decodeIfPresent(Int64.self, forKey: .playDate) ?? 0
        achievement = try container.decodeIfPresent(Int64.self, forKey: .achievement) ?? 0
        scoreRank = try container.decodeIfPresent(Int.self, forKey: .scoreRank) ?? 0
        deluxeScore = try container.decodeIfPresent(Int.self, forKey: .deluxeScore) ?? 0
        comboStatus = try container.decodeIfPresent(Int.self, forKey: .comboStatus) ?? 0
        syncStatus = try container.decodeIfPresent(Int.self, forKey: .syncStatus) ?? 0
    }
}

struct OtogameMusic: Decodable, Sendable {
    let musicID: String
    let name: String
    let isDeluxe: Bool
    let utageKanjiName: String?

    private enum CodingKeys: String, CodingKey {
        case musicID = "music_id"
        case name
        case isDeluxe = "is_deluxe"
        case utageKanjiName = "utage_kanji_name"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if let stringID = try? container.decode(String.self, forKey: .musicID) {
            musicID = stringID
        } else if let integerID = try? container.decode(Int64.self, forKey: .musicID) {
            musicID = String(integerID)
        } else {
            musicID = ""
        }
        name = try container.decodeIfPresent(String.self, forKey: .name) ?? ""
        isDeluxe = try container.decodeIfPresent(Bool.self, forKey: .isDeluxe) ?? false
        utageKanjiName = try container.decodeIfPresent(String.self, forKey: .utageKanjiName)
    }
}

struct OtogameLevelInfo: Decodable, Sendable {
    let difficulty: Int
    let level: Int

    static let empty = OtogameLevelInfo(difficulty: -1, level: 0)
}

File: android/app/src/main/java/org/rhythmeta/maimaid/core/data/OtogamePlaylogModels.kt
package org.rhythmeta.maimaid.core.data

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonNames
import kotlinx.serialization.json.jsonPrimitive

@OptIn(ExperimentalSerializationApi::class)
@Serializable
internal data class OtogamePlaylogResponse(
    val code: String = "",
    val message: String = "",
    val data: OtogamePlaylogPage = OtogamePlaylogPage(),
)

@Serializable
internal data class OtogamePlaylogPage(
    val data: List<OtogamePlaylog> = emptyList(),
    val pagination: OtogamePagination = OtogamePagination(),
)

@OptIn(ExperimentalSerializationApi::class)
@Serializable
internal data class OtogamePagination(
    val page: Int = 1,
    @SerialName("per_page")
    @JsonNames("perPage")
    val perPage: Int = 0,
    @SerialName("total_page")
    @JsonNames("totalPage")
    val totalPage: Int = 1,
)

@OptIn(ExperimentalSerializationApi::class)
@Serializable
internal data class OtogamePlaylog(
    val music: OtogameMusic = OtogameMusic(),
    val difficulty: Int = -1,
    @SerialName("level_info")
    @JsonNames("levelInfo")
    val levelInfo: OtogameLevelInfo = OtogameLevelInfo(),
    @SerialName("track_no")
    @JsonNames("trackNo")
    val trackNo: Int = 0,
    @SerialName("play_date")
    @JsonNames("playDate")
    val playDate: Long = 0,
    val achievement: Long = 0,
    @SerialName("score_rank")
    @JsonNames("scoreRank")
    val scoreRank: Int = 0,
    @SerialName("deluxe_score")
    @JsonNames("deluxeScore")
    val deluxeScore: Int = 0,
    @SerialName("combo_status")
    @JsonNames("comboStatus")
    val comboStatus: Int = 0,
    @SerialName("sync_status")
    @JsonNames("syncStatus")
    val syncStatus: Int = 0,
)

@OptIn(ExperimentalSerializationApi::class)
@Serializable
internal data class OtogameMusic(
    @SerialName("music_id")
    @JsonNames("musicId")
    @Serializable(with = OtogameMusicIdSerializer::class)
    val musicId: String = "",
    val name: String = "",
    @SerialName("is_deluxe")
    @JsonNames("isDeluxe")
    val isDeluxe: Boolean = false,
    @SerialName("utage_kanji_name")
    @JsonNames("utageKanjiName")
    val utageKanjiName: String? = null,
)

@Serializable
internal data class OtogameLevelInfo(
    val difficulty: Int = -1,
    val level: Int = 0,
)

private object OtogameMusicIdSerializer : KSerializer<String> {
    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("OtogameMusicId", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): String =
        (decoder as JsonDecoder).decodeJsonElement().jsonPrimitive.content

    override fun serialize(encoder: Encoder, value: String) {
        encoder.encodeString(value)
    }
}
